x/programs/wasmlanche/src/host.rs

```diff
@@ -0,0 +1,258 @@
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(feature = "test")]
+impl MockState {
+    pub fn new() -> Self {
+        Self {
+            state: hashbrown::HashMap::new(),
+        }
+    }
+
+    pub fn get(&self, key: &[u8]) -> HostPtr {
+        match self.state.get(key) {
+            Some(val) => {
+                let ptr = crate::memory::alloc(val.len());
+                unsafe {
+                    std::ptr::copy(val.as_ptr(), ptr.as_ptr().cast_mut(), val.len());
+                }
+                ptr
+            }
+            None => HostPtr::null(),
+        }
+    }
+}
+
+cfg_if! {
+    if #[cfg(feature = "test")] {
+        #[derive(Clone)]
+        #[cfg_attr(feature = "debug", derive(Debug))]
+        pub struct Accessor {
+            pub state: MockState,
+        }
+    } else {
+        #[derive(Clone)]
+        #[cfg_attr(feature = "debug", derive(Debug))]
+        pub struct Accessor;
+    }
+}
+
+#[cfg(feature = "test")]
+impl Default for Accessor {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(feature = "test")]
+impl Accessor {
+    pub fn new() -> Self {
+        Accessor {
+            state: MockState::new(),
+        }
+    }
+
+    pub fn deploy(&mut self, ptr: *const u8, len: usize) -> HostPtr {
+        let key = unsafe { std::slice::from_raw_parts(ptr, len) };
+        let key = [DEPLOY_PREFIX]
+            .iter()
+            .chain(key.iter())
+            .copied()
+            .collect::<Vec<u8>>();
+        let val = self.state.get(&key);
+
+        assert!(
+            !val.is_null(),
+            "Deploy function not mocked. Please mock the function call."
+        );
+
+        val
+    }
+
+    pub fn call_program(&self, ptr: *const u8, len: usize) -> HostPtr {
+        let key = unsafe { std::slice::from_raw_parts(ptr, len) };
+        let key = [CALL_FUNCTION_PREFIX]
+            .iter()
+            .chain(key.iter())
+            .copied()
+            .collect::<Vec<u8>>();
+        let val = self.state.get(&key);
+
+        assert!(
+            !val.is_null(),
+            "Call function not mocked. Please mock the function call."
+        );
+
+        val
+    }
+
+    pub fn get_balance(&self, ptr: *const u8, len: usize) -> HostPtr {
+        let key = unsafe { std::slice::from_raw_parts(ptr, len) };
+        // balance prefix + key
+        let key = [BALANCE_PREFIX]
+            .iter()
+            .chain(key.iter())
+            .copied()
+            .collect::<Vec<u8>>();
+
+        let host_ptr = self.state.get(&key);
+        assert!(
+            !host_ptr.is_null(),
+            "get_balance not mocked. Please mock the function call."
+        );
+
+        host_ptr
+    }
+
+    #[allow(clippy::unused_self)]
+    pub fn get_remaining_fuel(&self) -> HostPtr {
+        panic!("get_remaining_fuel not implemented in the test context");
+    }
+
+    pub fn send_value(&self, ptr: *const u8, len: usize) -> HostPtr {
```

Comment:
All of these methods are unsafe. We should be passing in the proper types and converting them to pointers inside these methods.

Reply:
hmm, we need to keep the function signatures the same for both compiler directives tho?

Reply:
nvm, see what ur saying will update